tests/services/test_sarses_generator.py

```diff
@@ -0,0 +1,225 @@
+# pylint: disable-all
+
+import logging
+import pytest
+from app.services.sarses_generator import generate_sarses
+from tests.factories.marketplace_factories import UserFactory
+from tests.factories.recommendation_factory import RecommendationFactory
+from tests.factories.user_action_factory import UserActionFactory
+from app.db.mongo_models import Sars
+from tests.fixtures.utils import mongo
+from app.reward_mapping import ua_to_reward_id
+from app.services.sarses_generator import (
+    RECOMMENDATION_PAGES_IDS,
+    _tree_collapse,
+    _get_clicked_services_and_reward,
+    _find_root_uas_before,
+)
+from app.db.mongo_models import UserAction
+
+LOGGER = logging.getLogger(__name__)
+
+
+def ruas2services(ruas):
+    return list(map(lambda rua: rua.source.root.service, ruas))
+
+
+class TestSarsesGenerator:
+    def test_tree_colapse(self, mongo):
+        # Zero level
+        root_ua = UserActionFactory(recommendation_root=True)
+
+        rewards = _tree_collapse(root_ua)
+        assert type(rewards) == list
+        assert len(rewards) == 1
+
+        # First level
+        ua0 = UserActionFactory(source__visit_id=root_ua.target.visit_id)
+        ua1 = UserActionFactory(source__visit_id=root_ua.target.visit_id)
+        ua2 = UserActionFactory(source__visit_id=root_ua.target.visit_id)
+        ua3 = UserActionFactory(source__visit_id=root_ua.target.visit_id)
```

Comment:
Just a syntax thing, but this could be shortened to something along these lines:

```
ua0, ua1, ua2, ua3 = UserActionFactory.create_batch(4, source__visit_id=root_ua.target.visit_id)
```